osu.Game/Overlays/BeatmapSet/LeaderboardModSelector.cs

```diff
@@ -0,0 +1,165 @@
+using System.Linq;
+using osu.Framework.Extensions.IEnumerableExtensions;
+using osu.Framework.Graphics.Sprites;
+
+namespace osu.Game.Overlays.BeatmapSet
+{
+    public class LeaderboardModSelector : CompositeDrawable
+    {
+        public readonly BindableList<Mod> SelectedMods = new BindableList<Mod>();
+        public readonly Bindable<RulesetInfo> Ruleset = new Bindable<RulesetInfo>();
+
+        private readonly FillFlowContainer<ModButton> modsContainer;
+
+        public LeaderboardModSelector()
+        {
+            AutoSizeAxes = Axes.Both;
+            InternalChild = modsContainer = new FillFlowContainer<ModButton>
+            {
+                Anchor = Anchor.Centre,
+                Origin = Anchor.Centre,
+                AutoSizeAxes = Axes.Both,
+                Direction = FillDirection.Full,
+                Spacing = new Vector2(4),
+            };
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            Ruleset.BindValueChanged(onRulesetChanged, true);
+        }
+
+        private void onRulesetChanged(ValueChangedEvent<RulesetInfo> ruleset)
+        {
+            SelectedMods.Clear();
+            modsContainer.Clear();
+
+            if (ruleset.NewValue == null)
+                return;
+
+            modsContainer.Add(new ModButton(new NoMod()));
+
+            ruleset.NewValue.CreateInstance().GetAllMods().ForEach(mod =>
+            {
+                if (mod.Ranked)
+                    modsContainer.Add(new ModButton(mod));
+            });
+
+            modsContainer.ForEach(button => button.OnSelectionChanged += selectionChanged);
+        }
+
+        private void selectionChanged(Mod mod, bool selected)
+        {
+            if (selected)
+                SelectedMods.Add(mod);
+            else
+                SelectedMods.Remove(mod);
+
+            if (!SelectedMods.Any() && !IsHovered)
+                highlightAll();
+        }
+
+        protected override bool OnHover(HoverEvent e)
+        {
+            if (!SelectedMods.Any())
+                modsContainer.ForEach(button =>
+                {
+                    if (!button.IsHovered)
+                        button.Highlighted.Value = false;
+                });
+
+            return base.OnHover(e);
+        }
+
+        protected override void OnHoverLost(HoverLostEvent e)
+        {
+            base.OnHoverLost(e);
+
+            if (!SelectedMods.Any())
+                highlightAll();
+        }
+
+        public void DeselectAll() => modsContainer.ForEach(mod => mod.Selected.Value = false);
+
+        private void highlightAll() => modsContainer.ForEach(mod => mod.Highlighted.Value = true);
+
+        private class ModButton : ModIcon
+        {
+            private const float mod_scale = 0.4f;
+            private const int duration = 200;
+
+            public readonly BindableBool Selected = new BindableBool();
+            public Action<Mod, bool> OnSelectionChanged;
```

Comment:
Supposed to be an event by its usage.
```suggestion
            public event Action<Mod, bool> OnSelectionChanged;
```

Reply:
Any big difference?